pkg/pool-weighted/contracts/WeightedMath.sol

```diff
@@ -414,40 +414,46 @@ library WeightedMath {
         return amountsOut;
     }
 
-    function _calcDueTokenProtocolSwapFeeAmount(
-        uint256 balance,
-        uint256 normalizedWeight,
+    function _calcDueProtocolSwapFeeBPTAmount(
+        uint256 totalSupply,
         uint256 previousInvariant,
         uint256 currentInvariant,
         uint256 protocolSwapFeePercentage
     ) internal pure returns (uint256) {
-        /*********************************************************************************
-        /*  protocolSwapFeePercentage * balanceToken * ( 1 - (previousInvariant / currentInvariant) ^ (1 / weightToken))
-        *********************************************************************************/
-
         if (currentInvariant <= previousInvariant) {
             // This shouldn't happen outside of rounding errors, but have this safeguard nonetheless to prevent the Pool
             // from entering a locked state in which joins and exits revert while computing accumulated swap fees.
             return 0;
         }
 
+        // Assuming the Pool is balanced and token weights have not changed, a growth of the invariant translates into
+        // proportional growth of all token balances. The protocol is due a percentage of that growth: more precisely,
+        // it is due `fee * (growth - 1) * balance / growth` for each token.
+        // We compute the amount of BPT to mint for the protocol that would allow it to proportionally exit the Pool and
+        // receive these balances. Note that the total BPT supply will increase when minting, so we need to account for
+        // this in order to compute the percentage of Pool ownership the protocol will have.
+
+        // The formula is:
+        //
+        //            supply * protocol fee * (growth - 1)
+        //                     ---------------------------
+        //                              growth
+        // toMint = --------------------------------------
+        //            1 - protocol fee * (growth - 1)
+        //                ---------------------------
+        //                          growth
+        //
```

Comment:
Another way to think of it (maybe clearer?) is like this:

Call Inv0 the original (last) invariant, and Inv1 the new invariant before a join/exit.
TS0 is the original totalSupply, and TS1 is the new totalSupply after minting to pay fees.

Consider a 50/50 pool with 0.5 WBTC and 5 WETH. Inv0 = 1.5811; TS0 = 3.1622
Swap until all balances go up 20%: we now have 0.6 WBTC and 6 WETH.
Inv1 = 1.8974: 20% higher than (= 1.2 times) Inv0

The ratio of invariants is the proportional amount of growth, with 1 = no change, 0 growth.
The growth percentage, then, is the amount of increase over 1:

                       Inv1
growthPct = ——  - 1 = 1.2 - 1 = 0.2 (20%)
                       Inv0

The bptShare due the protocol is then:

                                     growthPct
bptShare = feePct  *  ——————
                                   1 + growthPct

Intuitively, it’s the increase over the new total (= 1 + increase), since we’re calculating *after* the mint.
If the fee were 100% (1), the bptShare would be the entire increase.

The new total supply, then, would be TS0 / LPShare (= 1 - bptShare). 

                TS0
TS1 =  —————
           1 - bptShare

And the amount to mint would be simply the protocol’s share of the *post mint* supply:

toMint = TS1 * bptShare

This is identical math - it was just clearer to me expressed this way.